lighthouse-core/computed/bundle-analysis.js

```diff
@@ -0,0 +1,151 @@
+ * @param {TextSourceMap} map
+ * @param {string} content
+ */
+function computeGeneratedFileSizes(map, content) {
+  const lines = content.split('\n');
+  /** @type {Record<string, number>} */
+  const files = {};
+  const totalBytes = content.length;
+  let unmappedBytes = totalBytes;
+
+  // If the map + contents don't line up, return a result that
+  // denotes nothing is mapped.
+  const failureResult = {files: {}, unmappedBytes, totalBytes};
+
+  // @ts-ignore: This function is added in SDK.js
+  map.computeLastGeneratedColumns();
+
+  for (const mapping of map.mappings()) {
+    const source = mapping.sourceURL;
+    const lineNum = mapping.lineNumber;
+    const colNum = mapping.columnNumber;
+    // @ts-ignore: `lastColumnNumber` is not on types yet.
+    const lastColNum = /** @type {number=} */ (mapping.lastColumnNumber);
+
+    // Webpack sometimes emits null mappings.
+    // https://github.com/mozilla/source-map/pull/303
+    if (!source) continue;
+
+    // Lines are 1-based
+    const line = lines[lineNum];
+    if (line === null) {
+      log.error(`${map.compiledURL} mapping for line out of bounds: ${lineNum + 1}`);
+      return failureResult;
+    }
+
+    // Columns are 0-based
+    if (colNum >= line.length) {
+      log.error(`${map.compiledURL} mapping for column out of bounds: ${lineNum + 1}:${colNum}`);
+      return failureResult;
+    }
+
+    let mappingLength = 0;
+    if (lastColNum !== undefined) {
+      if (lastColNum >= line.length) {
+        // eslint-disable-next-line max-len
+        log.error(`${map.compiledURL} mapping for last column out of bounds: ${lineNum + 1}:${lastColNum}`);
+        return failureResult;
+      }
+      mappingLength = lastColNum - colNum;
+    } else {
+      // TODO Buffer.byteLength?
+      mappingLength = line.length - colNum;
+    }
+    files[source] = (files[source] || 0) + mappingLength;
+    unmappedBytes -= mappingLength;
+  }
+
+  return {
+    files,
+    unmappedBytes,
+    totalBytes,
+  };
+}
+
+class BundleAnalysis {
+  /**
+   * @param {Pick<LH.Artifacts, 'SourceMaps'|'ScriptElements'|'devtoolsLogs'>} artifacts
+   * @param {LH.Audit.Context} context
+   */
+  static async compute_(artifacts, context) {
+    const {SourceMaps, ScriptElements} = artifacts;
+    const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+    const networkRecords = await NetworkRecords.request(devtoolsLog, context);
+
+    /** @type {Bundle[]} */
+    const bundles = [];
+
+    // Collate map, script, and network record.
+    for (let mapIndex = 0; mapIndex < SourceMaps.length; mapIndex++) {
+      const SourceMap = SourceMaps[mapIndex];
+      if (!SourceMap.map) continue;
+      const {scriptUrl, map: rawMap} = SourceMap;
+
+      const scriptElement = ScriptElements.find(s => s.src === scriptUrl);
+      const networkRecord = networkRecords.find(r => r.url === scriptUrl);
+      if (!scriptElement) continue;
+
+      // Lazily generate expensive things.
```

Comment:
Do we want to avoid generating sizes for ALL scripts/maps? It seems reasonable, but I'd just like to verify this assumption because it's effort to handle that.

Reply:
If the answer is yes and if we want this to be lazy then really **this computed artifact should be per script**. something like

```js
const {map, sizes} = ScriptBundle.request({scriptElement, networkRecord}, context);
```

(presumably we don't need have a usecase of just the map without the sizes..)

Reply:
The intention is to lazily generate sizes for just the scripts as needed, which is what this current impl. + the `getters` do now. Additionally, it could be that some audits need a bundle's CDT map (but not the sizes).

I didn't realize the syntax you gave is possible - didn't know each computed artifact memoizes based on the first parameter. I understand it's a better API b/c is makes the lazy mechanism more obvious.

I'm against it mostly because it would require each audit to link a ScriptElement w/ a NetworkRecord themselves - which I think is cumbersome programming. I really like how this computed artifact does all that for you.

Reply:
Also, the ScriptBundle would need to be given the map, so it'd be more like `ScriptBundle.request({scriptElement, networkRecord, sourceMap})`

Reply:
To provide some data on why I think the optimization isn't necessary and maybe you can correct my ignorance... 
IIUC, [this line](https://github.com/GoogleChrome/lighthouse/pull/10064/files#diff-73ca4fab96b7e4e60bd197c25538f3aaR60) will already force every single map to be evaluated so why both lazily doing them?

Reply:
oof @ that LOC you linked to. If we wanted to be lazy, we'd have to do something like store the visualization data on the audit `context` and "subscribe" to a bundle's visualization data as needed in each audit.

> i.e. In what situations will Lighthouse, across all audits, skip computationally expensive source maps that were valid and how often do those occur?

Ignoring the bundle visualization audit, I'd expect the _size_ information to be not needed for all bundles, but the _cdt map_ would be need for all (b/c of duplication audit). So should definitely remove the laziness from the latter.

Let me run some numbers on calculating sizes. IIRC it is the most expensive bit, but I'll also collect timings for creating the map and for collecting the visualization data.

> Maybe we can create a helper for this then? This is annoying, but it's not difficult, and it's a pattern that happens often that other audits could benefit from. Also, in this scenario then it looks like the networkRecord is not really involved in the logic at all, so the API would just be ScriptBundle.request({scriptElement, sourceMap}) to be linked later.

What about a new computed artifact that computes the grouped script / map / network record - and the input of each one can be given to `BundleAnalysis.request` to get a single bundle?

```
const scriptData = await ScriptData.request(artifacts);
const bundle = await BundleAnalysis.request(scriptData);
// yay bundle.map (not lazy)
// perhaps bundle.calculateSizes() to denote cost (and we cache it to make it lazy)
```

Reply:
> Ignoring the bundle visualization audit, I'd expect the size information to be not needed for all bundles, but the cdt map would be need for all (b/c of duplication audit). So should definitely remove the laziness from the latter.


Ugh, but now I recall that I want `valid-source-maps` to link to each bundle visualization, which means that we need sizing info for everything. In that case, all this lazy optimization is moot.